service/iot-service-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/service/auth/TokenCredentialCacheTest.java

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package tests.unit.com.microsoft.azure.sdk.iot.service.auth;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+import com.microsoft.azure.sdk.iot.service.auth.TokenCredentialCache;
+import com.sun.scenario.effect.Offset;
+import mockit.Expectations;
+import mockit.Mocked;
+import org.joda.time.field.OffsetDateTimeField;
+import org.junit.Test;
+import reactor.core.publisher.Mono;
+
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+import static org.junit.Assert.assertEquals;
+
+public class TokenCredentialCacheTest
+{
+    @Mocked
+    TokenCredential mockTokenCredential;
+
+    @Mocked
+    AccessToken mockAccessToken;
+
+    @Mocked
+    AccessToken mockAccessToken2;
+
+    @Mocked
+    Mono<AccessToken> mockTokenTask;
+
+    @Test
+    public void tokenCredentialCachesToken()
+    {
+        TokenCredentialCache cache = new TokenCredentialCache(mockTokenCredential);
+
+        new Expectations()
+        {
+            {
+                mockTokenCredential.getToken((TokenRequestContext) any).block();
+                result = mockAccessToken;
+
+                mockAccessToken.getExpiresAt();
+                result = OffsetDateTime.MAX;
+            }
+        };
+
+        AccessToken accessToken = cache.getAccessToken();
+        AccessToken accessToken2 = cache.getAccessToken();
+
+        assertEquals(mockAccessToken, accessToken);
+        assertEquals(mockAccessToken, accessToken2);
+    }
+
+    @Test
+    public void tokenCredentialProactivelyRenewsToken()
+    {
+        TokenCredentialCache cache = new TokenCredentialCache(mockTokenCredential);
+
+        new Expectations()
+        {
+            {
+                mockTokenCredential.getToken((TokenRequestContext) any).block();
+                result = mockAccessToken;
+
+                mockAccessToken.getExpiresAt();
+                result = OffsetDateTime.MAX;
+            }
+        };
+
+        AccessToken accessToken = cache.getAccessToken();
+
+        // 8 minutes from the current time, should fit within the proactive renewal range
+        final long milliseconds = System.currentTimeMillis() + (8 * 60 * 1000);
+        new Expectations()
+        {
+            {
+                mockAccessToken.getExpiresAt();
+                result = Instant.ofEpochMilli(milliseconds).atOffset(ZoneOffset.UTC);
+
+                mockTokenCredential.getToken((TokenRequestContext) any).block();
```

Comment:
I might not be understanding the syntax here well but does this not mean you are setting the value for what should be returned on calling getToken? How will this test the logic? I think you should be testing the logic in isAccessTokenCloseToExpiry which return true or false to be sure your dateTime comparison logic is tested.

Reply:
The way I read this was:
Setup - If getToken is call, return mockAccessToken
Execute - Call getToken
So there will be no situation that these values will be different.

Reply:
This is mocking the TokenCredential that powers the class under test. This line means that the test expects the tokenCredential to get a token, and the result of that call is defined by the line after it.

Reply:
Mocking the AccessToken's ```getExpiresAt``` method is what is really helping to test the ```isAccessTokenCloseToExpiry``` method

Reply:
So how can you turn this into a test that might fail? If you change isAccessTokenCloseToExpiry from <= 9 to >9, will it fail?

Reply:
Ok I think I finally understand the syntax! Thanks for the explanation :)